homeassistant/components/rhasspy/stt.py

```diff
@@ -0,0 +1,197 @@
+
+
+class RhasspySTTProvider(Provider):
+    """Rhasspy speech to text provider."""
+
+    def __init__(self, hass, conf):
+        """Create Rhasspy speech to text provider."""
+        self.hass = hass
+        self.config = conf
+
+        # URL to stream microphone audio
+        self.speech_url = conf.get(CONF_SPEECH_URL, None)
+        self.headers = {"Content-Type": "audio/wav"}
+
+    async def async_process_audio_stream(
+        self, metadata: SpeechMetadata, stream: aiohttp.StreamReader
+    ) -> SpeechResult:
+        """Process an audio stream to STT service.
+
+        Only streaming of content are allow!
+        """
+        if self.speech_url is None:
+            self.speech_url = get_speech_url(self.hass)
+
+        _LOGGER.debug("Receiving audio")
+        text_result = ""
+
+        try:
+            # First chunk is a WAV header (no frames)
+            header_chunk = True
+            with io.BytesIO() as wav_io:
+                wav_file = wave.open(wav_io, "wb")
+                async for audio_chunk, _ in stream.iter_chunks():
+                    if header_chunk:
+                        # Extract WAV information
+                        header_chunk = False
+                        with io.BytesIO(audio_chunk) as header_io:
+                            with wave.open(header_io) as header_file:
+                                wav_file.setnchannels(header_file.getnchannels())
+                                wav_file.setsampwidth(header_file.getsampwidth())
+                                wav_file.setframerate(header_file.getframerate())
+                    else:
+                        # Everything after first chunk is audio data.
+                        # Add to in-memory WAV file.
+                        wav_file.writeframes(audio_chunk)
+
+                wav_file.close()
+                wav_data = wav_io.getvalue()
+                _LOGGER.debug("Received %s byte(s)", len(wav_data))
+
+            # POST to Rhasspy server
+            session = async_get_clientsession(self.hass)
+            async with session.post(
+                self.speech_url, headers=self.headers, data=wav_data
+            ) as res:
+                res.raise_for_status()
+                text_result = await res.text()
+                _LOGGER.debug(text_result)
+
+                return SpeechResult(text=text_result, result=SpeechResultState.SUCCESS)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("async_process_audio_stream")
+
+        return SpeechResult(text="", result=SpeechResultState.ERROR)
+
+    # -------------------------------------------------------------------------
+
+    @property
+    def supported_languages(self) -> List[str]:
+        """Return a list of supported languages."""
+        return SUPPORT_LANGUAGES
+
+    @property
+    def supported_formats(self) -> List[AudioFormats]:
+        """Return a list of supported formats."""
+        return [AudioFormats.WAV]
```

Comment:
You set this to wave, so all software need send an wave header and you don't need deal with this

Reply:
I'd rather not deal with the WAV header, but the only AudioFormats I saw were WAV and OGG. If you can add RAW, I can set everything to 16Khz 16-bit mono and stream directly to Rhasspy.

Reply:
Streaming now to Rhasspy, and just dropping the first chunk (WAV header).